src/insertItem.ts

```diff
@@ -0,0 +1,449 @@
+        }
+    }
+    return false;
+}
+
+export function getQuickPickItem(label: string): vscode.QuickPickItem {
+    return { label: label };
+}
+
+export function getItemTypeQuickPicks(): QuickPickItem<TemplateSectionType>[] {
+    let items: QuickPickItem<TemplateSectionType>[] = [];
+    items.push(new QuickPickItem<TemplateSectionType>("Function", TemplateSectionType.Functions, "Insert a function"));
+    items.push(new QuickPickItem<TemplateSectionType>("Output", TemplateSectionType.Outputs, "Insert an output"));
+    items.push(new QuickPickItem<TemplateSectionType>("Parameter", TemplateSectionType.Parameters, "Insert a parameter"));
+    items.push(new QuickPickItem<TemplateSectionType>("Resource", TemplateSectionType.Resources, "Insert a resource"));
+    items.push(new QuickPickItem<TemplateSectionType>("Variable", TemplateSectionType.Variables, "Insert a variable"));
+    return items;
+}
+
+export class InsertItem {
+    private ui: IAzureUserInput;
+
+    constructor(ui: IAzureUserInput) {
+        this.ui = ui;
+    }
+
+    public async insertItem(template: DeploymentTemplate | undefined, sectionType: TemplateSectionType, textEditor: vscode.TextEditor): Promise<void> {
+        if (!template) {
+            return;
+        }
+        ext.outputChannel.appendLine("Insert item");
+        switch (sectionType) {
+            case TemplateSectionType.Functions:
+                if (await this.insertFunction(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the output of the function.");
+                }
+                break;
+            case TemplateSectionType.Outputs:
+                if (await this.insertOutput(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the the value of the output.");
+                }
+                break;
+            case TemplateSectionType.Parameters:
+                if (await this.insertParameter(template, textEditor)) {
+                    vscode.window.showInformationMessage("Done inserting parameter.");
+                }
+                break;
+            case TemplateSectionType.Resources:
+                if (await this.insertResource(template, textEditor)) {
+                    vscode.window.showInformationMessage("Press TAB to move between the tab stops.");
+                }
+                break;
+            case TemplateSectionType.Variables:
+                if (await this.insertVariable(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the the value of the variable.");
+                }
+                break;
+            default:
+                vscode.window.showWarningMessage("Unknown insert item type!");
+                return;
+        }
+    }
+
+    private getTemplateObjectPart(template: DeploymentTemplate, templatePart: string): Json.ObjectValue | undefined {
+        let part = this.getTemplatePart(template, templatePart);
```

Comment:
suggestion: return this.getTemplatePart(template, templatePart)?.asObjectValue;

(as recently added the asObjectValue/asArrayValu/etc properties)

Reply:
Fixed.